cmd/sippy/component_readiness.go

```diff
@@ -0,0 +1,170 @@
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"gopkg.in/yaml.v3"
+)
+
+type ComponentReadinessFlags struct {
+	GoogleCloudFlags *flags.GoogleCloudFlags
+	BigQueryFlags    *flags.BigQueryFlags
+	CacheFlags       *flags.CacheFlags
+
+	Config      string
+	LogLevel    string
+	ListenAddr  string
+	MetricsAddr string
+	RedisURL    string
+}
+
+func NewComponentReadinessCommand() *cobra.Command {
+	f := &ComponentReadinessFlags{
+		LogLevel:    "info",
+		ListenAddr:  ":8080",
+		MetricsAddr: ":2112",
+
+		GoogleCloudFlags: flags.NewGoogleCloudFlags(),
+		BigQueryFlags:    flags.NewBigQueryFlags(),
+		CacheFlags:       flags.NewCacheFlags(),
+	}
+
+	cmd := &cobra.Command{
+		Use: "component-readiness",
+
+		Run: func(cmd *cobra.Command, arguments []string) {
+
+			if err := f.Validate(); err != nil {
+				log.WithError(err).Fatalf("error validation options")
+			}
+			if err := f.Run(); err != nil {
+				log.WithError(err).Fatalf("error running command")
+			}
+		},
+	}
+
+	f.BindFlags(cmd.Flags())
+
+	return cmd
+}
+
+func (f *ComponentReadinessFlags) BindFlags(flags *pflag.FlagSet) {
+	f.CacheFlags.BindFlags(flags)
+	f.BigQueryFlags.BindFlags(flags)
+	f.GoogleCloudFlags.BindFlags(flags)
+	flags.StringVar(&f.LogLevel, "log-level", f.LogLevel, "Log level (trace,debug,info,warn,error) (default info)")
+}
+
+func (f *ComponentReadinessFlags) Validate() error {
+	return nil
+}
+
+func (f *ComponentReadinessFlags) Run() error { //nolint:gocyclo
+	// Set log level
+	level, err := log.ParseLevel(f.LogLevel)
+	if err != nil {
+		log.WithError(err).Fatal("Cannot parse log-level")
+	}
+	log.SetLevel(level)
+
+	// Add some millisecond precision to log timestamps, useful for debugging performance.
+	formatter := new(log.TextFormatter)
+	formatter.TimestampFormat = "2006-01-02T15:04:05.999Z07:00"
+	formatter.FullTimestamp = true
+	formatter.DisableColors = false
+	log.SetFormatter(formatter)
+
+	log.Debug("debug logging enabled")
+	sippyConfig := v1.SippyConfig{}
+	if f.Config == "" {
+		sippyConfig.Prow = v1.ProwConfig{
+			URL: "https://prow.ci.openshift.org/prowjobs.js",
+		}
+	} else {
+		data, err := os.ReadFile(f.Config)
+		if err != nil {
+			log.WithError(err).Fatalf("could not load config")
+		}
+		if err := yaml.Unmarshal(data, &sippyConfig); err != nil {
+			log.WithError(err).Fatalf("could not unmarshal config")
+		}
+	}
+
+	return f.runServerMode()
+}
+
+func (f *ComponentReadinessFlags) runServerMode() error {
+	var err error
+
+	webRoot, err := fs.Sub(resources.SippyNG, "sippy-ng/build")
+	if err != nil {
+		log.WithError(err).Fatal("could not load frontend")
+	}
+
+	cacheClient, err := f.CacheFlags.GetCacheClient()
+	if err != nil {
+		return errors.WithMessage(err, "couldn't get cache client")
+	}
+
+	var bigQueryClient *bigquery.Client
+	var gcsClient *storage.Client
+	if f.GoogleCloudFlags.ServiceAccountCredentialFile != "" {
```

Comment:
I'd like to push this logic into the GoogleCloudFlags type, so we can be consistent with how to get output from the input.